File: backend/app/schemas/produce.py
from typing import Literal, Optional
from pydantic import BaseModel, ConfigDict, field_validator


class ProduceBase(BaseModel):
    name: str
    description: Optional[str] = None
    category: Optional[Literal["Vegetables", "Fruits"]] = None
    price: int = 0
    unit: Literal["kg", "lbs"] = "kg"
    unit_quantity: float = 1.0
    stock_quantity: int = 0
    image_url: Optional[str] = None


class ProduceCreate(ProduceBase):
    pass


class ProduceUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    category: Optional[Literal["Vegetables", "Fruits"]] = None
    price: Optional[int] = None
    unit: Optional[Literal["kg", "lbs"]] = None
    unit_quantity: Optional[float] = None
    stock_quantity: Optional[int] = None
    image_url: Optional[str] = None

    @field_validator("category", mode="before")
    @classmethod
    def coerce_empty_category(cls, v: object) -> object:
        return None if v == "" else v


class ProduceResponse(ProduceBase):
    id: int
    merchant_id: int

    model_config = ConfigDict(from_attributes=True)

File: frontend/carabao-app/util/api/produce.ts
import { apiClient } from './client';

export type Produce = {
  id: number;
  merchant_id: number;
  name: string | null;
  description: string | null;
  category: 'Vegetables' | 'Fruits' | null;
  price: number;
  unit: 'kg' | 'lbs';
  unit_quantity: number;
  stock_quantity: number;
  image_url: string | null;
};

export type ProduceCreatePayload = {
  name: string;
  description?: string | null;
  category?: 'Vegetables' | 'Fruits' | null;
  price: number;
  unit?: 'kg' | 'lbs';
  unit_quantity?: number;
  stock_quantity?: number;
  image_url?: string | null;
};

export type ProduceUpdatePayload = {
  name?: string | null;
  description?: string | null;
  category?: 'Vegetables' | 'Fruits' | null;
  price?: number | null;
  unit?: 'kg' | 'lbs' | null;
  unit_quantity?: number | null;
  stock_quantity?: number | null;
  image_url?: string | null;
};

export const fetchMerchantProduce = async (merchantId: number): Promise<Produce[]> => {
  const response = await apiClient.get<Produce[]>(`/merchants/${merchantId}/produce`);
  return response.data;
};

export const createProduce = async (payload: ProduceCreatePayload): Promise<Produce> => {
  const response = await apiClient.post<Produce>('/produce/', payload);
  return response.data;
};

export const updateProduce = async (produceId: number, payload: ProduceUpdatePayload): Promise<Produce> => {
  const response = await apiClient.patch<Produce>(`/produce/${produceId}`, payload);
  return response.data;
};

export const deleteProduce = async (produceId: number): Promise<void> => {
  await apiClient.delete(`/produce/${produceId}`);
};

export const uploadProduceImage = async (produceId: number, file: File): Promise<Produce> => {
  const formData = new FormData();
  formData.append('image', file);
  const response = await apiClient.post<Produce>(`/produce/${produceId}/image`, formData);
  return response.data;
};
